### PR TITLE
paginated addressbook

### PR DESCRIPTION
**Pull request by jbrit**
## Checklist

- [x] The PR modified the backend to modify the address book endpoint

### CODE REVIEW

**Review by yabirgb — COMMENTED** [review]
the new address in the evm detection is expected to fail. For the other error I'm not sure what happened... are you rebased?

File: rotkehlchen/db/addressbook.py [hunk 1]
```diff
@@ -48,45 +49,28 @@ def write_ctx(self, book_type: AddressbookType) -> Generator['DBCursor', None, N
     def get_addressbook_entries(
             self,
             cursor: 'DBCursor',
-            optional_chain_addresses: Optional[list[OptionalChainAddress]] = None,
-    ) -> list[AddressbookEntry]:
+            filter_query: Optional['AddressbookFilterQuery'] = None,
```
Comment (yabirgb) [inline rotkehlchen/db/addressbook.py:52]: why make it optional? It complicates the logic and irc in other similar functions we always provide it, even if is empty

File: docs/api.rst [hunk 2]
```diff
@@ -5020,17 +5020,17 @@ Exporting History Events
    :statuscode 409: No user is logged in or failure at event export.
    :statuscode 500: Internal rotki error
 
-.. http:patch:: /api/(version)/history/events/export
+.. http:put:: /api/(version)/history/events/export
 
-   Doing a PATCH on this endpoint with the given filter parameters will download a csv with all history events matching the filter. All arguments are optional. If no filter is used all the events will be downloaded.
+   Doing a PUT on this endpoint with the given filter parameters will download a csv with all history events matching the filter. All arguments are optional. If no filter is used all the events will be downloaded.
```
Comment (yabirgb) [inline docs/api.rst:5025]: is this change intentional? was it wrong?

**Review by yabirgb — COMMENTED** [review]
One issue and a few questions

File: rotkehlchen/db/addressbook.py (hunk 1 above)
Comment (jbrit) [reply rotkehlchen/db/addressbook.py:52]: You mean, I shouldn't provide a default value of None?, or I should not make it optional at all? i.e. users should create an empty filter?

File: docs/api.rst (hunk 2 above)
Comment (jbrit) [reply docs/api.rst:5025]: It's intentional, I'm pretty sure this was on develop before

File: docs/api.rst [hunk 3]
```diff
@@ -5020,17 +5020,17 @@ Exporting History Events
    :statuscode 409: No user is logged in or failure at event export.
    :statuscode 500: Internal rotki error
 
-.. http:patch:: /api/(version)/history/events/export
+.. http:put:: /api/(version)/history/events/export
```
Comment (LefterisJP) [inline docs/api.rst:5023]: This is a correct change but there is one more thing off. The name of the function here should be `put`:

https://github.com/rotki/rotki/blob/2b358c20a052729238f79c24733d88bc8448e6d9/rotkehlchen/api/v1/resources.py#L2971

Please do it since you are changing this here in the docs

File: rotkehlchen/db/filtering.py [hunk 4]
```diff
@@ -580,6 +582,25 @@ class DBMultiIntegerFilter(DBMultiValueFilter[int]):
...
+class DBOptionalChainAddressesFilter(DBFilter):
+    """Filter the address column by a selection of optional chain addresses"""
+    optional_chain_addresses: Optional[list[OptionalChainAddress]]
+
+    def prepare(self) -> tuple[list[str], list[str]]:
+        query_filters = []
+        bindings: list[Union[str, ChecksumEvmAddress]] = []
+        if self.optional_chain_addresses is not None:
+            for optional_chain_address in self.optional_chain_addresses:
+                query_part = 'address = ?'
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:595]: This is a weird way of doing it. Instead I want you to use already existing functionality for filtering for a chain of addresses.

Check `DBTransactionsPendingDecodingFilter` in this same file: https://github.com/rotki/rotki/blob/2b358c20a052729238f79c24733d88bc8448e6d9/rotkehlchen/db/filtering.py#L174

And instead abstract that logic and move it here. And then use this `DBOptionalChainAddressesFilter` in `DBTransactionsPendingDecodingFilter` through inheritance as is done in other places in this file.

File: rotkehlchen/tests/api/test_addressbook.py [hunk 5]
```diff
@@ -56,7 +57,7 @@ def test_get_addressbook(
         ),
     )
     result = assert_proper_response_with_result(response)
-    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result]
+    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result['entries']]  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_addressbook.py:60]: here assert for entries found and total too

File: rotkehlchen/tests/api/test_addressbook.py [hunk 6]
```diff
@@ -73,9 +74,8 @@ def test_get_addressbook(
         },
     )
     result = assert_proper_response_with_result(response=response)
-    expected_entries = generated_entries[0:3]
-    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result]
-    assert set(deserialized_entries) == set(expected_entries)
+    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result['entries']]  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_addressbook.py:77]: same here

File: rotkehlchen/tests/api/test_addressbook.py [hunk 7]
```diff
@@ -86,14 +86,72 @@ def test_get_addressbook(
...
+                    'blockchain': SupportedBlockchain.OPTIMISM.serialize(),
+                },
             ],
         },
     )
     result = assert_proper_response_with_result(response=response)
-    expected_entries = [generated_entries[1]]
-    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result]
-    assert set(deserialized_entries) == set(expected_entries)
+    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result['entries']]  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_addressbook.py:97]: same here

File: rotkehlchen/tests/api/test_addressbook.py [hunk 8]
```diff
@@ -86,14 +86,72 @@ def test_get_addressbook(
...
+            'offset': 1,
+            'limit': 1,
+            'blockchain': SupportedBlockchain.ETHEREUM.serialize(),
+        },
+    )
+    result = assert_proper_response_with_result(response=response)
+    deserialized_entries = [AddressbookEntry.deserialize(data=raw_entry) for raw_entry in result['entries']]  # noqa: E501
+    assert set(deserialized_entries) == {generated_entries[2]}
+    assert result['entries_found'] == 2
+    assert result['entries_total'] == 5
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_addressbook.py:154]: After this you should also add a filter by a list of addresses. Make sure to have at least two tests. One for 1 address and 1 for 2 addresses.

**Review by LefterisJP — COMMENTED** [review]
Left some comments

File: rotkehlchen/db/filtering.py (hunk 4 above)
Comment (jbrit) [reply rotkehlchen/db/filtering.py:595]: They two filters are very different, and check different fields

Comment (LefterisJP) [reply rotkehlchen/db/filtering.py:595]: Discussed this in a call and this is indeed two different things as this is using:

```python
class OptionalChainAddress(NamedTuple):
    address: ChecksumAddress
    blockchain: Optional[SupportedBlockchain]
    ```

My comment is invalid.

**Review by LefterisJP — APPROVED** [review]
LGTM. Thank you Ajibolah
